include/mapbox/variant.hpp

```diff
@@ -178,34 +178,16 @@ struct value_traits
     using target_type = typename std::tuple_element<tindex, std::tuple<void, Types...>>::type;
 };
 
-template <typename T, typename R = void>
-struct enable_if_type
-{
-    using type = R;
-};
-
-template <typename F, typename V, typename Enable = void>
+template <typename F, typename V>
 struct result_of_unary_visit
 {
-    using type = typename std::result_of<F(V&)>::type;
+    using type = decltype(::std::declval<F>()(::std::declval<V>()));
 };
 
 template <typename F, typename V>
-struct result_of_unary_visit<F, V, typename enable_if_type<typename F::result_type>::type>
-{
-    using type = typename F::result_type;
-};
-
-template <typename F, typename V, typename Enable = void>
 struct result_of_binary_visit
 {
-    using type = typename std::result_of<F(V&, V&)>::type;
-};
-
-template <typename F, typename V>
-struct result_of_binary_visit<F, V, typename enable_if_type<typename F::result_type>::type>
-{
-    using type = typename F::result_type;
+    using type = decltype(::std::declval<F>()(::std::declval<V>(), (void)::std::declval<V>()));
```

Comment:
`lhs ,(void) rhs` required to prevent user overloaded `operator,` to trigger

Reply:
Old comment, but I think it's worth rectifying. A comma in a function-call argument list is a delimiter, not an operator, hence overloaded `operator,` does not come into play. And then you're asking what would be the return value of the call if the second argument was of type `void` -- the answer will be something like "invalid use of void".